dataflow/gemma-flex-template/README.md

```diff
@@ -0,0 +1,179 @@
+        self._env_vars = {}
+
+    def load_model(self) -> GemmaForCausalLM:
+        """Loads and initializes a model for processing."""
+        torch.set_default_dtype(self._model_config.get_dtype())
+        model = GemmaForCausalLM(self._model_config)
+        model.load_weights(self._checkpoint_path)
+        model = model.to(self._device).eval()
+        return model
+```
+
+### Formatting DoFn
+
+The output from a keyed model handler is a tuple of the form `(key, PredictionResult)`. To format that output into a string before sending it to the answer Pub/Sub topic, use an extra `DoFn`.
+
+```py
+class FormatOutput(beam.DoFn):
+  def process(self, element, *args, **kwargs):
+    yield "Key : {key}, Input: {input}, Output: {output}".format(key=element[0], input=element[1].example, output=element[1].inference)
+```
+
+## Build the Flex Template
+Run the following code from the directory to build the Dataflow flex template.
+
+- Replace `$TEMPLATE_FILE` with a Google Cloud Storage path written as a JSON file.
+- Set `SDK_CONTAINER_IMAGE` to the name of the Docker image created previously.
+- `$PROJECT` is the Google Cloud project that you created previously. 
+
+```sh
+gcloud dataflow flex-template build $TEMPLATE_FILE \
+  --image $SDK_CONTAINER_IMAGE \
+  --sdk-language "PYTHON" \
+  --metadata-file metadata.json \
+  --project $PROJECT
+```
+
+## Start the pipeline
+To start the Dataflow streaming job, run the following code from the directory. Replace `$TEMPLATE_FILE`, `$REGION`, `$GCS_BUCKET`, `$INPUT_SUBSCRIPTION`, `$OUTPUT_TOPIC`, `$SDK_CONTAINER_IMAGE`, and `$PROJECT` with the Google Cloud project resources you created previously. It might take as much as 15 minutes for the worker to start up and to begin accepting messages from the input Pub/Sub topic. 
+
+```sh
+gcloud dataflow flex-template run "gemma-flex-`date +%Y%m%d-%H%M%S`" \
+  --template-file-gcs-location $TEMPLATE_FILE \
+  --region $REGION \
+  --staging-location $GCS_BUCKET \
+  --parameters messages_subscription=$INPUT_SUBSCRIPTION \
+  --parameters responses_topic=$OUTPUT_TOPIC \
+  --parameters device="GPU" \
+  --parameters sdk_container_image=$SDK_CONTAINER_IMAGE \
+  --parameters dataflow_service_options="worker_accelerator=type:nvidia-l4;count:1;install-nvidia-driver:5xx" \
+  --project $PROJECT \
+  --worker-machine-type "g2-standard-4"
+```
+
+## Send a prompt to the model and check the response
+
+In the Google Cloud console, navigate to the Pub/Sub topics page, and then select your input topic. On the **Messages** tab, click **Publish Message**. Add a message for the Dataflow job to pick up and pass through the model. The Dataflow job outputs the response to the Pub/Sub sink topic. To check the response from the model, you can manually pull messages from the destination topic. For more information, see [Publish messages](https://cloud.google.com/pubsub/docs/publisher#publish-messages) in the Google Cloud documentation.
```

Comment:
I did not know what to send as a message. I would suggest some sample input. 

I also had to add retention to the sink to generate a pull subscription to be able to pull responses to the job. I understand that in practice there might be follow on uses of the sink topic, but for the purposes of this example, there were additional steps needed to check for success.

Reply:
added a sample